### PR TITLE
MINOR: add code quality checks to checkstyle.xml. Also add suppressions

### PR DESCRIPTION
**Pull request by dguy**
Add code quality/complexity checks to checkstyle

### CODE REVIEW

**Comment by dguy** [comment]
@ewencp @ijuma - i've been through the java modules and added checkstyle. Adjusted the defaults a bit to be more lenient (so i don't scare too many people). Also added a bunch of suppressions (i didn't want to change any code in doing this).

**Comment by ijuma** [comment]
Seems like you need to adjust the suppressions @dguy.

**Comment by dguy** [comment]
@ijuma - yep :-(

**Review by ijuma — COMMENTED** [review]
Thanks for the PR. I think it's good that we are looking to do more checks automatically.

One concern I have is that we have a number of exclusions that rely on line numbers. That looks fragile and people are likely to run into problems when doing unrelated changes. Could we start with using file-based exclusions and then attempting to fix whole files at a time?

I'd be interested in other people's opinions on this.

**Comment by dguy** [comment]
@ijuma yes i tend to agree (i've already been bitten by it when i submitted this PR). My only concern is that it continues to allow people to practice bad behaviour in those closes. But i'm ok with changing it. The alternative is to use  `//CHECKSTYLE:OFF` and `//CHECKSTYLE:ON` in the code, but i didn't really want to do that.

**Comment by ijuma** [comment]
@dguy yes, I agree that it's not ideal. But a step forward still since most classes are not excluded.

**Comment by ijuma** [comment]
Also, can we take the chance to upgrade to a newer version of checkstyle? It should be a matter of setting `toolVersion = '6.19'` (checkstyle `7.x` requires Java 8, worth adding a comment stating this so that we know to bump it when we upgrade to Java 8).

**Comment by dguy** [comment]
sure i'll upgrade it.

**Comment by dguy** [comment]
@ijuma updated to 6.19. There were a few minor changes i need to make to get checkstyle to pass the `NoWhitespace...` rules must be more strict. Also removed the line numbers as discussed

File: checkstyle/suppressions.xml [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+    <suppress checks="ParameterNumber"
+              files="NetworkClient.java"
+              lines="132"/>
+    <suppress checks="ParameterNumber"
+              files="KafkaConsumer.java"
+              lines="722"/>
+    <suppress checks="ParameterNumber"
+              files="ConsumerCoordinator.java"
+              lines="91"/>
```
Comment (ijuma) [inline checkstyle/suppressions.xml]: 3 cases above where we have `lines` still. And a fourth one in `WorkerSourceTask`. Did you leave them intentionally?

Comment (dguy) [reply checkstyle/suppressions.xml]: nope. thought i'd removed them all

**Review by ijuma — APPROVED** [review]
Thanks for the updates, LGTM. What do you think @ewencp @hachikuji @guozhangwang?

**Comment by guozhangwang** [comment]
LGTM! Thanks @dguy .

**Comment by ewencp** [comment]
LGTM

**Comment by hachikuji** [comment]
Seems like a good idea. I might quibble on the using the parameter length restriction though. The alternative is often to use a builder which is a style we haven't fully adopted in the codebase. But perhaps it's fine as a default since we can add additional suppressions if the usage really has no reasonable alternative.

One nit: there's a bunch of unnecessary whitespace in the suppressions file. Maybe we could clean that up on merging?

**Comment by ijuma** [comment]
I'll merge to trunk and fix the unnecessary whitespace in the suppressions file.

**Pull request closed without merging**